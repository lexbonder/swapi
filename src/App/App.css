nav {
  display: flex;
  justify-content: space-around;

  width: 70%;
  margin: 20px auto;
}

h1 {
  color: red;
  font-size: 24px;
  width: 40%;
  margin: auto;
  text-align: center;
  border: 3px solid red;
  border-radius: 10px;
  padding: 10px
}

h2 {
  text-align: center;
  text-decoration: underline;
  color: yellow;
}

section {
  width: 70%;
  margin: 0 auto 20px;

}

section div {
  display: grid;
  grid-template-columns: 1fr 1fr;
  border-bottom: 1px solid white;
}

img {
  display: block;
  width: 55%;
  margin: auto;
}

#loading {
  text-decoration: none;
}

@media screen and (max-width: 1010px) {
  nav {
    flex-wrap: wrap;
    max-width: 380px;
  }

  section {
    width: fit-content;
  }

  section div {
    grid-template-columns: 1fr;
  }
}

@media screen and (max-width: 540px) {
  body {
    background-size: auto;
  }

  nav {
    justify-content: center;
    max-width: 320px;
  }
}