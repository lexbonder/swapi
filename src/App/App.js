import React, { Component } from 'react';
import Character from '../Character/Character';
import apiCall from '../apiCall';
import { characters } from '../characters.json';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      films: []
    };

  }
  
  handleClick = (event) => {
    const { id } = event.target;
    const clicked = characters.find(char => char.name === id);
    this.getFilmUrls(clicked.url);
  }

  getFilmUrls = async (url) => {
    const { films } = await apiCall(url);
    this.setState({ films });
  }

  renderCharBtns = () => {
    return characters.map(({ name }) => {
      return <Character key={name} name={name} click={this.handleClick} />;
    });
  }

  render() {
    return (
      <div className="App">
        <nav>
          {this.renderCharBtns()}
        </nav>
        <main>
          
        </main>
      </div>
    );
  }
}

export default App;
