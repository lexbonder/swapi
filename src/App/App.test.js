/* eslint-disable */
import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import { shallow } from "enzyme";

describe("App", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a default state", () => {
    const expectedState = {
      films: [],
      error: "",
      loading: false
    };
    expect(wrapper.state()).toEqual(expectedState);
  });

  describe("handleClick", () => {
    it("should call setFilmData with the url of the chosen character", () => {
      const mockEvent = { target: { id: "Luke Skywalker" } };
      wrapper.instance().setFilmData = jest.fn();

      wrapper.instance().handleClick(mockEvent);

      expect(wrapper.instance().setFilmData).toHaveBeenCalledWith(
        "https://swapi.co/api/people/1/"
      );
    });

    it("should call reset", () => {
      const mockEvent = { target: { id: "Luke Skywalker" } };
      wrapper.instance().reset = jest.fn();

      wrapper.instance().handleClick(mockEvent);

      expect(wrapper.instance().reset).toHaveBeenCalled();
    });

    it("should set loading state to true", () => {
      const mockEvent = { target: { id: "Luke Skywalker" } };
      wrapper.instance().handleClick(mockEvent);

      expect(wrapper.state().loading).toEqual(true);
    });
  });

  describe("setFilmData", () => {
    it("should set the state of films with an array of films", async () => {
      wrapper.instance().dataCleaner.getFilmInfo = jest.fn().mockImplementationOnce(() => Promise.resolve([
            {
              title: "The Empire Strikes Back",
              date: "Saturday, May 17th 1980"
            }
          ]));
      const expected = [
        {
          title: "The Empire Strikes Back",
          date: "Saturday, May 17th 1980"
        }
      ];

      await wrapper.instance().setFilmData("nonsense");
      expect(wrapper.state().films).toEqual(expected);
    });

    it('should set an error in state if retreval fails', async () => {
      wrapper.instance().dataCleaner.getFilmInfo = jest.fn()
        .mockImplementationOnce(() => Promise.reject(
          'Much failure'
          ))
      await wrapper.instance().setFilmData('nonsense')
      expect(wrapper.state().error).toEqual('Much failure')
    })

    it('should set state of loading to false', () => {
      expect(wrapper.state().loading).toEqual(false);
    })
  });

  describe('reset', () => {
    it('should set the films array and error back to default', () => {
      wrapper.setState({films: [1,2,3,4], error:'full of sadness'});
      expect(wrapper.state().films).toEqual([1,2,3,4]);
      expect(wrapper.state().error).toEqual("full of sadness");
      wrapper.instance().reset()
      expect(wrapper.state().films).toEqual([])
      expect(wrapper.state().error).toEqual('')
    })
  })

  describe('renderCharBtns', () => {
    it('should match the snapshot', () => {
      expect(wrapper.instance().renderCharBtns()).toMatchSnapshot()
    })
  })

  describe('renderResults', () => {
    it('should match the snapshot when there is an error in state', () => {
      wrapper.setState({error: 'So much sadness'});
      expect(wrapper.instance().renderResults()).toMatchSnapshot();
    })
    
    it('should match the snapshot where there are results to show', () => {
      wrapper.setState({
        films: [{ 
          title: 'Empire Strikes Back',
          date: "Saturday, May 17th 1980"
        }]
      });
      expect(wrapper.instance().renderResults()).toMatchSnapshot();
    })
  })
});
