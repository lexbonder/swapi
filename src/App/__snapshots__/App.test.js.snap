// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renderCharBtns should match the snapshot 1`] = `
Array [
  <Character
    click={[Function]}
    name="Luke Skywalker"
/>,
  <Character
    click={[Function]}
    name="Darth Vader"
/>,
  <Character
    click={[Function]}
    name="Obi-wan Kenobi"
/>,
  <Character
    click={[Function]}
    name="R2-D2"
/>,
]
`;

exports[`App renderResults should match the snapshot when there is an error in state 1`] = `
<h1>
  So much sadness
</h1>
`;

exports[`App renderResults should match the snapshot where there are results to show 1`] = `
<div>
  <h2>
    Movies Featuring 
    
    :
  </h2>
  <section>
    <div>
      <h3>
        Title
      </h3>
      <h3>
        Release Date
      </h3>
    </div>
    <Film
      date="Saturday, May 17th 1980"
      title="Empire Strikes Back"
    />
  </section>
</div>
`;

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <nav>
          <Character
                    click={[Function]}
                    name="Luke Skywalker"
          />
          <Character
                    click={[Function]}
                    name="Darth Vader"
          />
          <Character
                    click={[Function]}
                    name="Obi-wan Kenobi"
          />
          <Character
                    click={[Function]}
                    name="R2-D2"
          />
</nav>,
        false,
        <main />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Character
              click={[Function]}
              name="Luke Skywalker"
/>,
            <Character
              click={[Function]}
              name="Darth Vader"
/>,
            <Character
              click={[Function]}
              name="Obi-wan Kenobi"
/>,
            <Character
              click={[Function]}
              name="R2-D2"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Luke Skywalker",
            "nodeType": "function",
            "props": Object {
              "click": [Function],
              "name": "Luke Skywalker",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Darth Vader",
            "nodeType": "function",
            "props": Object {
              "click": [Function],
              "name": "Darth Vader",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Obi-wan Kenobi",
            "nodeType": "function",
            "props": Object {
              "click": [Function],
              "name": "Obi-wan Kenobi",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "R2-D2",
            "nodeType": "function",
            "props": Object {
              "click": [Function],
              "name": "R2-D2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "nav",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <nav>
            <Character
                        click={[Function]}
                        name="Luke Skywalker"
            />
            <Character
                        click={[Function]}
                        name="Darth Vader"
            />
            <Character
                        click={[Function]}
                        name="Obi-wan Kenobi"
            />
            <Character
                        click={[Function]}
                        name="R2-D2"
            />
</nav>,
          false,
          <main />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Character
                click={[Function]}
                name="Luke Skywalker"
/>,
              <Character
                click={[Function]}
                name="Darth Vader"
/>,
              <Character
                click={[Function]}
                name="Obi-wan Kenobi"
/>,
              <Character
                click={[Function]}
                name="R2-D2"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Luke Skywalker",
              "nodeType": "function",
              "props": Object {
                "click": [Function],
                "name": "Luke Skywalker",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Darth Vader",
              "nodeType": "function",
              "props": Object {
                "click": [Function],
                "name": "Darth Vader",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Obi-wan Kenobi",
              "nodeType": "function",
              "props": Object {
                "click": [Function],
                "name": "Obi-wan Kenobi",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "R2-D2",
              "nodeType": "function",
              "props": Object {
                "click": [Function],
                "name": "R2-D2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "nav",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
